File: src/shapes/triangle.js
// Triangle shape geometry calculations
export const triangleGeometry = {
  // Generate SVG path for triangle
  generatePath: (base, height, angle = 60) => {
    // Calculate vertices based on base, height, and top angle
    const angleRad = (angle * Math.PI) / 180;
    
    // Calculate the top vertex position
    const topX = base / 2;
    const topY = 0;
    
    // Calculate bottom vertices
    const bottomLeftX = 0;
    const bottomLeftY = height;
    const bottomRightX = base;
    const bottomRightY = height;
    
    return `M ${topX} ${topY} L ${bottomRightX} ${bottomRightY} L ${bottomLeftX} ${bottomLeftY} Z`;
  },

  // Calculate bounding box
  getBoundingBox: (base, height, angle = 60) => {
    return {
      minX: 0,
      minY: 0,
      maxX: base,
      maxY: height,
      width: base,
      height: height
    };
  },

  // Calculate area
  getArea: (base, height) => {
    return (base * height) / 2;
  },

  // Calculate perimeter
  getPerimeter: (base, height, angle = 60) => {
    const angleRad = (angle * Math.PI) / 180;
    
    // Calculate the length of the two equal sides (for isosceles triangle)
    const sideLength = Math.sqrt(Math.pow(base / 2, 2) + Math.pow(height, 2));
    
    return base + (2 * sideLength);
  },

  // Get center point (centroid)
  getCenter: (base, height) => {
    return {
      x: base / 3,
      y: height / 3
    };
  },

  // Get vertices
  getVertices: (base, height, angle = 60) => {
    return [
      { x: base / 2, y: 0 },        // Top vertex
      { x: base, y: height },        // Bottom-right
      { x: 0, y: height }            // Bottom-left
    ];
  },

  // Get dimension points for labels
  getDimensionPoints: (base, height, angle = 60) => {
    return {
      base: {
        start: { x: 0, y: height + 20 },
        end: { x: base, y: height + 20 },
        label: { x: base / 2, y: height + 35 }
      },
      height: {
        start: { x: base + 20, y: 0 },
        end: { x: base + 20, y: height },
        label: { x: base + 35, y: height / 2 }
      },
      angle: {
        vertex: { x: base / 2, y: 0 },
        label: { x: base / 2, y: -20 }
      }
    };
  },

  // Calculate side lengths
  getSideLengths: (base, height, angle = 60) => {
    const angleRad = (angle * Math.PI) / 180;
    const sideLength = Math.sqrt(Math.pow(base / 2, 2) + Math.pow(height, 2));
    
    return {
      base: base,
      leftSide: sideLength,
      rightSide: sideLength
    };
  },

  // Validate parameters
  validateParameters: (base, height, angle = 60) => {
    const errors = [];
    
    if (base <= 0) errors.push('Base must be greater than 0');
    if (height <= 0) errors.push('Height must be greater than 0');
    if (angle <= 0 || angle >= 180) errors.push('Angle must be between 0 and 180 degrees');
    if (base > 10000) errors.push('Base must be less than 10000mm');
    if (height > 10000) errors.push('Height must be less than 10000mm');
    
    // Check if triangle is geometrically possible (relaxed validation)
    // For isosceles triangle, any positive height should work
    if (height > base * 2) {
      errors.push('Height is too large for the given base');
    }
    
    return {
      isValid: errors.length === 0,
      errors
    };
  },

  // Generate different triangle types
  generateRightTriangle: (base, height) => {
    return {
      path: `M 0 0 L ${base} 0 L 0 ${height} Z`,
      vertices: [
        { x: 0, y: 0 },
        { x: base, y: 0 },
        { x: 0, y: height }
      ]
    };
  },

  generateEquilateralTriangle: (sideLength) => {
    const height = (sideLength * Math.sqrt(3)) / 2;
    return {
      path: `M ${sideLength / 2} 0 L ${sideLength} ${height} L 0 ${height} Z`,
      vertices: [
        { x: sideLength / 2, y: 0 },
        { x: sideLength, y: height },
        { x: 0, y: height }
      ]
    };
  }
};

export default triangleGeometry;

File: src-tauri/src/shapes/triangle.rs
use crate::shapes::{BoundingBox, Dimension, DimensionOrientation, Point, ShapeGeometry, Transform, ValidationError, ShapeParameters};

pub struct TriangleGeometry;

impl ShapeGeometry for TriangleGeometry {
    fn generate_path(&self, params: &ShapeParameters) -> String {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        let angle = params.angle.unwrap_or(60.0);
        
        // Calculate triangle vertices
        let center_x = base / 2.0;
        let top_y = 0.0;
        let bottom_y = height;
        
        // Calculate top vertex based on angle
        let _angle_rad = (angle * std::f64::consts::PI) / 180.0;
        let top_x = center_x;
        
        format!(
            "M {} {} L {} {} L {} {} Z",
            top_x, top_y,                    // Top vertex
            0.0, bottom_y,                   // Bottom-left
            base, bottom_y                   // Bottom-right
        )
    }

    fn get_bounding_box(&self, params: &ShapeParameters) -> BoundingBox {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        
        BoundingBox {
            min_x: 0.0,
            min_y: 0.0,
            max_x: base,
            max_y: height,
            width: base,
            height,
        }
    }

    fn get_area(&self, params: &ShapeParameters) -> f64 {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        0.5 * base * height
    }

    fn get_perimeter(&self, params: &ShapeParameters) -> f64 {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        
        // Calculate the length of the two equal sides
        let side_length = (base * base / 4.0 + height * height).sqrt();
        base + 2.0 * side_length
    }

    fn get_center(&self, params: &ShapeParameters) -> Point {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        
        // Centroid of a triangle is at 1/3 of the height from the base
        Point {
            x: base / 2.0,
            y: height / 3.0,
        }
    }

    fn get_vertices(&self, params: &ShapeParameters) -> Vec<Point> {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        
        vec![
            Point { x: base / 2.0, y: 0.0 },      // Top vertex
            Point { x: 0.0, y: height },         // Bottom-left
            Point { x: base, y: height },        // Bottom-right
        ]
    }

    fn validate_parameters(&self, params: &ShapeParameters) -> ValidationError {
        let mut errors = Vec::new();
        
        if let Some(base) = params.base {
            if base <= 0.0 {
                errors.push("Base must be greater than 0".to_string());
            }
            if base > 10000.0 {
                errors.push("Base must be less than 10000mm".to_string());
            }
        } else {
            errors.push("Base is required".to_string());
        }
        
        if let Some(height) = params.height {
            if height <= 0.0 {
                errors.push("Height must be greater than 0".to_string());
            }
            if height > 10000.0 {
                errors.push("Height must be less than 10000mm".to_string());
            }
        } else {
            errors.push("Height is required".to_string());
        }
        
        if let Some(angle) = params.angle {
            if angle <= 0.0 || angle >= 180.0 {
                errors.push("Angle must be between 0 and 180 degrees".to_string());
            }
        }
        
        ValidationError {
            is_valid: errors.is_empty(),
            errors,
        }
    }

    fn get_dimensions(&self, params: &ShapeParameters, render_offset: &Point, transform: &Transform) -> Vec<Dimension> {
        let base = params.base.unwrap_or(100.0);
        let height = params.height.unwrap_or(100.0);
        let offset = 20.0;
        
        // Create base dimensions (before transformation)
        let base_dimensions = vec![
            // Base dimension (horizontal at bottom)
            Dimension {
                start_point: Point { x: base *0.13 + render_offset.x, y: height +10.0 + render_offset.y },
                end_point: Point { x: base *0.87 + render_offset.x, y: height + 10.0 + render_offset.y },
                text_position: Point { x: base / 2.0 + render_offset.x, y: height + offset + 5.0 + render_offset.y },
                value: base,
                label: format!("{:.0}mm", base),
                orientation: DimensionOrientation::Horizontal,
            },
            // Height dimension (vertical from base to top, positioned inside triangle)
            Dimension {
                start_point: Point { x: base * 0.5 + render_offset.x, y: height -offset * 0.5 + render_offset.y },
                end_point: Point { x: base * 0.5 + render_offset.x, y: 10.0 + render_offset.y },
                text_position: Point { x: base * 0.5 + 15.0 + render_offset.x, y: height / 2.0 + render_offset.y },
                value: height ,
                label: format!("{:.0}mm", height ),
                orientation: DimensionOrientation::Vertical,
            },
        ];
        
        // Apply rotation transformation if needed
        if transform.rotation != 0.0 {
            let center = self.get_center(params);
            let adjusted_center = Point {
                x: center.x + render_offset.x,
                y: center.y + render_offset.y,
            };
            
            base_dimensions.into_iter().map(|mut dim| {
                dim.start_point = self.transform_point(&dim.start_point, &adjusted_center, transform);
                dim.end_point = self.transform_point(&dim.end_point, &adjusted_center, transform);
                dim.text_position = self.transform_point(&dim.text_position, &adjusted_center, transform);
                
                // Update dimension orientation based on rotation
                let rotation_degrees = transform.rotation;
                let normalized_rotation = rotation_degrees % 360.0;
                
                if (normalized_rotation >= 45.0 && normalized_rotation < 135.0) || 
                   (normalized_rotation >= 225.0 && normalized_rotation < 315.0) {
                    // Swap orientations for 90° and 270° rotations
                    match dim.orientation {
                        DimensionOrientation::Horizontal => dim.orientation = DimensionOrientation::Vertical,
                        DimensionOrientation::Vertical => dim.orientation = DimensionOrientation::Horizontal,
                        _ => {}
                    }
                }
                
                dim
            }).collect()
        } else {
            base_dimensions
        }
    }
    
    fn transform_point(&self, point: &Point, center: &Point, transform: &Transform) -> Point {
        let mut x = point.x;
        let mut y = point.y;

        // Apply rotation
        if transform.rotation != 0.0 {
            let angle_rad = (transform.rotation * std::f64::consts::PI) / 180.0;
            let cos_a = angle_rad.cos();
            let sin_a = angle_rad.sin();
            
            let rel_x = x - center.x;
            let rel_y = y - center.y;
            
            x = rel_x * cos_a - rel_y * sin_a + center.x;
            y = rel_x * sin_a + rel_y * cos_a + center.y;
        }

        Point { x, y }
    }

    fn get_rotation_center(&self, params: &ShapeParameters) -> Point {
        // For triangle, use the same center as the shape (centroid)
        self.get_center(params)
    }
}
